### PR TITLE
Add override exception for interface functions.

### LINKED ISSUE #8281: Do not require override for functions from interfaces
The override keyword is required when implementing a function from a parent interface.

```solidity
interface I {
    function foo() external;
}

contract C is I {
    function foo() external override { // Does not compile without override
    }
}
```

This doesn't sound right to me. The function is not overriding anything, it's simply filling in a missing abstract function.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #8281

### CODE REVIEW

**Comment by chriseth** [comment]
This seems to be a bit more complicated, because the override graph is not properly created of the "override" specifier is missing. The function that overrides the interface function is not considered overriding it as far as the graph is concerned.

**Comment by chriseth** [comment]
Since it turned out to be more complicated, I will not work on this in the next days.

**Comment by Marenz** [comment]
I'll be taking over this PR :)

File: Changelog.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ### 0.8.8 (unreleased)
 
 Language Features:
+ * Inheritance: A function that overrides a single interface function does not need the ``override`` specifier.
```
Comment (hrkrshnn) [inline Changelog.md]:
```suggestion
 * Inheritance: A function that overrides a only single interface function does not require the ``override`` specifier.
```

?

File: libsolidity/analysis/OverrideChecker.cpp [hunk 2]
```diff
@@ -86,10 +86,16 @@ struct OverrideGraph
...
+		bool addedEdge = false;
+		for (auto const& baseFunction: _function.baseFunctions())
+			if (_function.overrides() || baseFunction.contract().isInterface())
+			{
 				addEdge(currentNode, visit(baseFunction));
-		else
+				addedEdge = true;
+			}
+
+		if (!addedEdge)
```
Comment (hrkrshnn) [inline libsolidity/analysis/OverrideChecker.cpp]: Can you perhaps explain this change as a comment?

Comment (Marenz) [reply libsolidity/analysis/OverrideChecker.cpp]: Basically, before we only considered whether the function itself has the `override` keyword, now we also alternatively check if the base function is in an interface. For that we have to do the check inside the loop, though.

**Comment by chriseth** [comment]
Please also update the documentation.

File: libsolidity/analysis/OverrideChecker.cpp [hunk 3]
```diff
@@ -86,10 +86,18 @@ struct OverrideGraph
...
-		if (_function.overrides())
-			for (auto const& baseFunction: _function.baseFunctions())
+
+		bool addedEdge = false;
+		for (auto const& baseFunction: _function.baseFunctions())
+			// Add edges for overrides.
+			// Functions implementing an interface don't need the override specifier.
+			if (_function.overrides() || baseFunction.contract().isInterface())
+			{
 				addEdge(currentNode, visit(baseFunction));
```
Comment (chriseth) [inline libsolidity/analysis/OverrideChecker.cpp:92]: What if we just do this unconditionally?

File: test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol [hunk 4]
```diff
@@ -0,0 +1,10 @@
+interface I {
+	function f() external;
+}
+interface J {
+	function f() external;
+}
+contract C is I, J {
+	function f() external override (I, J) {}
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol]: The test is only meaningful if you do not supply `override (I, J)` here, isn't it?

File: test/libsolidity/syntaxTests/inheritance/override/interface_exception/abstract_needed.sol [hunk 5]
```diff
@@ -0,0 +1,7 @@
+abstract contract I {
+	function f() external virtual;
+}
+contract C is I {
+	function f() external override {}
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inheritance/override/interface_exception/abstract_needed.sol]: The test is only meaningful if you do not have `override` here, isn't it?

File: test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol (hunk 4 above)
Comment (Marenz) [reply test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol]: This PR didn't add the support for leaving out override with multiple parameters. Should it be added?

File: test/libsolidity/syntaxTests/inheritance/override/interface_exception/abstract_needed.sol (hunk 5 above)
Comment (Marenz) [reply test/libsolidity/syntaxTests/inheritance/override/interface_exception/abstract_needed.sol]: We aren't checking for abstract, only for interface though?

File: test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol]: No, I'm saying that you are not testing the property with the test.

File: test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+	function g() external;
+	function h() external;
+}
+contract C is I, J {
+	function f() external {}
+	function g() external override {}
+	function h() external override(I) {}
+}
+// ----
+// TypeError 4327: (198-222): Function needs to specify overridden contracts "I" and "J".
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inheritance/override/interface_exception/diamond_needed.sol:17]: Made some changes here.

File: docs/contracts/inheritance.rst [hunk 7]
```diff
@@ -303,6 +303,13 @@ contracts can no longer change the behaviour of that function.
   outside of interfaces. In interfaces, all functions are
   automatically considered ``virtual``.
 
+.. note::
+
+  Starting from Solidity 0.8.7, the ``override`` keyword is not
```
Comment (chriseth) [inline docs/contracts/inheritance.rst]: Added this note.

**Review by Marenz — DISMISSED** [review]

File: docs/contracts/inheritance.rst (hunk 7 above)
Comment (chriseth) [reply docs/contracts/inheritance.rst]: Version is wrong.

**Pull request merged**